### DISCUSSION TITLE
Best way to handle multiple-version mods?

### DISCUSSION THREAD

**Discussion by AnOpenSauceDev** [FabricMC/fabric-api#3205] [Mod Dev Support] [answered]
### Description

When developing mods for multiple versions of minecraft, a change in a class used by minecraft will inevitably cause issues/crashes. Is there a safe way to either bundle multiple mod versions in one JAR, or should I just maintain multiple branches of a mod?

**Comment by apple502j** [comment]
@AnOpenSauceDev People usually maintain multiple branches. You probably only maintain two or three branches at most, and it keeps the filesize smaller.

**Reply by AnOpenSauceDev** [reply]
That's what i usually do, but if it's not possible to handle for specific versions, i guess i have to stick to this method.

**Reply by AnAwesomGuy** [reply, marked as answer]
[reflection](https://github.com/AnAwesomGuy/stone-shears/blob/1.19.3-1.20.1/src/main/java/net/anawesomguy/stone_shears/StoneShears.java#L24-L29) :)
i highly doubt that using reflection is good practice but you can if you want to.

**Reply by AnOpenSauceDev** [reply]
Reflection is definitely a dark programming art, but technically answers this Q&A.

**Reply by caoimhebyrne** [reply]
I know this discussion is old, but reflection really isn't the best solution here, there's no guarantee that it'll work properly, and it's quite ugly to say the least.

I agree with apple502j, multiple branches if you just want to support the most recent X versions is the way to go.

**Comment by AnOpenSauceDev** [comment]
I found out something pretty interesting, you can totally package different JARs of your mod in `fabric.mod.json`, then check for the version to run. It's very slow to develop, but you can use [preprocessor](https://github.com/ReplayMod/preprocessor) to speed it up a bit. Overall it could be an interesting concept to use for anyone interested.
